templates/etcd_cluster/etcd_cluster.py

```diff
@@ -0,0 +1,274 @@
+from copy import deepcopy
+import gevent
+from jumpscale import j
+
+from zerorobot.template.base import TemplateBase
+from zerorobot.template.state import StateCheckError
+from zerorobot.service_collection import ServiceNotFoundError
+
+
+ETCD_TEMPLATE_UID = 'github.com/threefoldtech/0-templates/etcd/0.0.1'
+ZT_TEMPLATE_UID = 'github.com/threefoldtech/0-templates/zt_client/0.0.1'
+
+
+class EtcdCluster(TemplateBase):
+
+    version = '0.0.1'
+    template_name = "etcd_cluster"
+
+    def __init__(self, name=None, guid=None, data=None):
+        super().__init__(name=name, guid=guid, data=data)
+        self.add_delete_callback(self.uninstall)
+        self.recurring_action('_monitor', 30)  # every 30 seconds
+        self.recurring_action('_ensure_etcds_connections', 300)
+
+        self._robots = {}
+
+    def validate(self):
+        self.state.delete('status', 'running')
+        for nic in self.data['nics']:
+            if nic['type'] == 'zerotier':
+                break
+        else:
+            raise ValueError('Service must contain at least one zerotier nic')
+
+        if not self.data['farmerIyoOrg']:
+            raise ValueError('Invalid value for farmerIyoOrg')
+
+        self.data['password'] = self.data['password'] if self.data['password'] else j.data.idgenerator.generateXCharID(10)
+        self.data['token'] = self.data['token'] if self.data['token'] else j.data.idgenerator.generateXCharID(10)
```

Comment:
we could use the service guid here, since this token is supposed to identify the cluster itself